### ISSUE TITLE
SMB backend - response error: The contacted server does not support the indicated part of the DFS namespace.

### ISSUE THREAD

**Issue by tooptoop4** [rclone/rclone#6515]
#### What is the problem you are having with rclone?
SMB commands giving error

#### What is your rclone version (output from `rclone version`)
```
rclone v1.60.0
- os/version: debian 11.5 (64 bit)
- os/kernel: 4.14.294-220.533.amzn2.x86_64 (x86_64)
- os/type: linux
- os/arch: amd64
- go/version: go1.19.2
- go/linking: static
- go/tags: none
```

#### Which OS you are using and how many bits (e.g. Windows 7, 64 bit)
amznlinux2, 64bit

#### Which cloud storage system are you using? (e.g. Google Drive)
SMB

#### The command you were trying to run (e.g. `rclone copy /tmp remote:tmp`)
```rclone-linux-amd64/rclone ls 'smb:Shared/my/folders/here/' -vvv```

#### A log from the command with the `-vv` flag (e.g. output from `rclone -vv copy /tmp remote:tmp`)

```
root@hos:/# rclone-linux-amd64/rclone ls 'smb:Shared/my/folders/here/' -vvv
2022/10/24 02:09:22 DEBUG : rclone: Version "v1.60.0" starting with parameters ["rclone-linux-amd64/rclone" "ls" "smb:Shared/my/folders/here/" "-vvv"]
2022/10/24 02:09:22 DEBUG : Creating backend with remote "smb:Shared/my/folders/here/"
2022/10/24 02:09:22 DEBUG : Using config file from "/root/.config/rclone/rclone.conf"
2022/10/24 02:09:22 DEBUG : fs cache: renaming cache item "smb:Shared/my/folders/here/" to be canonical "smb:Shared/my/folders/here"
2022/10/24 02:09:22 ERROR : : error listing: open my\folders\here: response error: The contacted server does not support the indicated part of the DFS namespace.
2022/10/24 02:09:22 DEBUG : 5 go routines active
2022/10/24 02:09:22 DEBUG : smb://myuser@myhost:445/Shared/my/folders/here: Closing 1 unused connections
2022/10/24 02:09:22 Failed to ls with 2 errors: last error was: open my\folders\here: response error: The contacted server does not support the indicated part of the DFS namespace.
```

#### rclone.conf
```
[smb]
type = smb
host = myhost
user = myuser
pass = myencpw
domain = MYDOMAIN
```

note: have also raised https://github.com/hirochachacha/go-smb2/issues/69

https://github.com/siemens/GoScans/blob/48c07101ed3cf1815928b1809f2462d89318eb29/smb/smb_windows.go seems to have some code for handling DFS

more info on DFS:
https://wiki.samba.org/index.php/Distributed_File_System_(DFS)
https://www.samba.org/samba/docs/using_samba/ch08.html
https://learn.microsoft.com/en-us/windows-server/storage/dfs-namespaces/dfs-overview

**Comment by animosity22** [comment]
Can you please share:

rclone version output complete
rclone.conf any passwords redacted

Not sure how rclone relates to SMB offhand so if you can share some details, that would be great.

**Comment by tooptoop4** [comment]
@animosity22 updated
@Lesmiscore cc

**Comment by animosity22** [comment]
That rclone.conf doesn't seem valid or something is missing. There isn't an option for type=smb in rclone.

**Comment by tooptoop4** [comment]
@animosity22 https://github.com/rclone/rclone/commit/2aa264b33cb098f37eaeea43a87feec36b391ad0#diff-5e7c8d38b491c85f25c6bc51b64a8ef768c9f647bfccc7f8b2334d15194e751a added it?

**Comment by animosity22** [comment]
Ah, there we go. Thanks - I had my test machine on 1.59.2 and didn't realize it wasn't updated to 1.60. 

What's the other side connecting to?

**Comment by tooptoop4** [comment]
windows shared drive as seen in 
![](https://user-images.githubusercontent.com/33283496/197432590-5d0f5fdd-41df-4c19-a958-ed5d8d3920ba.png)

**Comment by Lesmiscore** [comment]
Please tell me how to do that in a consumer version of Windows (more specifically, Windows 11 Education) or the equivalent is possible in Samba?

**Comment by dblood** [comment]
I suspect I was encountering this same issue.  When connecting to the DFS endpoint I got this error message, but then I connected directly to the linked server and was able to list hidden shares only.  Because the target server is a cluster, the TargetSPN would need to be set:
https://github.com/rclone/rclone/blob/7db1c506f261698a536c343de39b628fc82e13bb/backend/smb/connpool.go#L39

Available arguments are:
https://github.com/hirochachacha/go-smb2/blob/f071e13222d669bd071ff798608e85107d679e09/initiator.go#L26

Unfortunately I don't have enough time to write a go client using hirochachacha/go-smb2 to verify the exact expected settings and to submit a patch to rclone.

**Comment by ncw** [comment]
It would be easy to add a config parameter for TargetSPN @dblood 

If I do that could you work out what it needs to be and write some docs I can paste in the code?

**Comment by dblood** [comment]
Though that would work for me while I bypass DFS, I think it would cause problems for DFS SMB endpoints where the DFS namespace is different from the target.  i.e. \\mycompany.com\internal points to \\server1\files\internal.

The ls of \\mycompany.com, \\server1 and \\xxx.xxx.xxx.xxx (server1's ip) may all be different, even if it is the same physical machine.  The SPN is what differentiates which shares are available but should match the requested.  

See the libraries comment:
https://github.com/hirochachacha/go-smb2/blob/f071e13222d669bd071ff798608e85107d679e09/internal/ntlm/client.go#L25

I *think* adding the following to configuration would work in most cases (but it would need to be tested)
`TargetSPN: "cifs/"+f.opt.Host+":"+f.opt.Port,`

**Comment by ncw** [comment]
Have a go with this @dblood 

[v1.62.0-beta.6711.4eda35406.fix-6515-smb-targetspn](https://beta.rclone.org/branch/fix-6515-smb-targetspn/v1.62.0-beta.6711.4eda35406.fix-6515-smb-targetspn/) on branch [fix-6515-smb-targetspn](https://github.com/rclone/rclone/tree/fix-6515-smb-targetspn) (uploaded in 15-30 mins)

I made the TargetSPN configurable with the `--smb-spn` flag or the `spn` config parameter.

Can you have a go with it and see if you can use it to fix your problem.

Here are the docs I came up with - any suggestions for improvements there would be apprecated!

#### --smb-spn

Service principal name.

Rclone presents this name to the server. Some servers use this as further
authentication, and it often needs to be set for clusters. For example:

    cifs/remotehost:1020

Leave blank if not sure.


Properties:

- Config:      spn
- Env Var:     RCLONE_SMB_SPN
- Type:        string
- Required:    false

**Comment by dblood** [comment]
I've managed to use the branch, and successfully authenticate with a spn.  Unfortunately the shares are still the hosts shares instead of the clusters.  

After reviewing code more, I believe the issue is in hirochachacha/go-smb2 and how all hostnames are converted to ip address instead of hostname through session.addr.  https://github.com/hirochachacha/go-smb2/blob/master/client.go#L72

\\xxx.xxx.xxx.xxx\shares are different from \\cluster\shares but the list shares function will only return \\xxx.xxx.xxx.xxx\shares

I've included some information for debugging SPNs for the future below to help anyone else looking into similar issues in the future (but SPN didn't resolve this issue).

To list the spn's for a host, you can use the following command from windows:
`setspn -l [host]`

The port is optional, so the targetSPN configured will probably be only `cifs/hostname`

**Comment by ncw** [comment]
@dblood I see what you mean. Do you want to make an upstream issue about this on the go-smb2 repo (put a link to this issue in if you do).

**Comment by dblood** [comment]
After searching more... I found https://github.com/hirochachacha/go-smb2/issues/40 and https://github.com/rclone/rclone/issues/6672 which is the same root issue I think.
@ncw you've already commented on those issues but is just waiting for the upstream library to make a decision.  If that fixes this ticket as well it might solve several open issues.

**Comment by dblood** [comment]
I've only dabbled in go so don't think I can put together an upstream PR and test it with current schedule.

**Comment by ncw** [comment]
I've merged the spn config variable to master now which means it will be in [the latest beta](https://beta.rclone.org/) in 15-30 minutes and released in v1.62 just in case it helps someone later.
